src/OpenTelemetry.Instrumentation.Wcf/Implementation/InstrumentedDuplexSessionChannel.cs

```diff
@@ -0,0 +1,188 @@
+// <copyright file="InstrumentedDuplexSessionChannel.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using System.Threading;
+
+namespace OpenTelemetry.Instrumentation.Wcf.Implementation;
+
+internal class InstrumentedDuplexSessionChannel : InstrumentedChannel, IDuplexSessionChannel
+{
+    private readonly TimeSpan telemetryTimeout;
+
+    public InstrumentedDuplexSessionChannel(IDuplexSessionChannel inner, TimeSpan telemetryTimeout)
+        : base(inner)
+    {
+        this.telemetryTimeout = telemetryTimeout;
+    }
+
+    public EndpointAddress LocalAddress => this.Inner.LocalAddress;
+
+    public EndpointAddress RemoteAddress => this.Inner.RemoteAddress;
+
+    public Uri Via => this.Inner.Via;
+
+    public IDuplexSession Session => this.Inner.Session;
+
+    private new IDuplexSessionChannel Inner { get => (IDuplexSessionChannel)base.Inner; }
+
+    public void Send(Message message)
+    {
+        this.SendInternal(message, _ => this.Inner.Send(message));
+    }
+
+    public void Send(Message message, TimeSpan timeout)
```

Comment:
This timeout is what you should be using for the telemetry tracking timeout. WCF channels created with a `ChannelFactory<TChannel>` shouldn't use the overload which has no timeout, but it's possible there's some edge cases where it doesn't (e.g. when used as an inner channel when using things like WS-Federation). To cover your basis, here's how you get the timeout when `Send(Message message)` is called. In `TelemetryBindingElement.BuildChannelFactory`, you need to capture the default send timeout from `BindingContext.Binding.SendTimeout` and pass that through to your instrumented channel factory implementations. From there, you need to pass the default send timeout to the Channel that you create. Use this as the timeout that you would pass to SendInternal.

Reply:
I just realized you already pass this value through. You would use this.telemetryTimeout when Send is called without a timeout, and you would use the timeout value pass in when Send is called with a timeout.

Reply:
Hah sweet, thanks....seems obvious now that you point it out 😄